service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/bucket/CopyGcsBucketDefinitionStep.java

```diff
@@ -0,0 +1,130 @@
+        Optional.ofNullable(
+                inputParameters.get(ControlledResourceKeys.DESTINATION_BUCKET_NAME, String.class))
+            .orElseGet(this::randomBucketName);
+    final UUID destinationWorkspaceId =
+        inputParameters.get(ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+    // get creation parameters together from working map and input map
+    // build a create flight
+    ControlledGcsBucketResource destinationBucket =
+        new ControlledGcsBucketResource(
+            destinationWorkspaceId,
+            UUID.randomUUID(),
+            resourceName,
+            description,
+            sourceBucket.getCloningInstructions(),
+            sourceBucket.getAssignedUser().orElse(null),
+            sourceBucket.getAccessScope(),
+            sourceBucket.getManagedBy(),
+            bucketName);
+    final ApiGcpGcsBucketCreationParameters sourceCreationParameters =
+        workingMap.get(
+            ControlledResourceKeys.CREATION_PARAMETERS, ApiGcpGcsBucketCreationParameters.class);
+    final List<ControlledResourceIamRole> iamRoles = getIamRoles(sourceBucket.getAccessScope());
+    final ControlledGcsBucketResource clonedBucket =
+        controlledResourceService.createBucket(
+            destinationBucket, sourceCreationParameters, iamRoles, userRequest);
+    final ApiCreatedControlledGcpGcsBucket apiCreatedBucket =
+        new ApiCreatedControlledGcpGcsBucket()
+            .gcpBucket(clonedBucket.toApiResource())
+            .resourceId(destinationBucket.getResourceId());
+    final ApiClonedControlledGcpGcsBucket apiBucketResult =
+        new ApiClonedControlledGcpGcsBucket()
+            .bucket(apiCreatedBucket)
+            .sourceWorkspaceId(sourceBucket.getWorkspaceId())
+            .sourceResourceId(sourceBucket.getResourceId());
+    workingMap.put(JobMapKeys.RESPONSE.getKeyName(), apiBucketResult);
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    final ControlledGcsBucketResource clonedBucket =
+        flightContext
+            .getWorkingMap()
+            .get(
+                ControlledResourceKeys.CLONED_RESOURCE_DEFINITION,
+                ControlledGcsBucketResource.class);
+    controlledResourceService.deleteControlledResourceSync(
+        clonedBucket.getWorkspaceId(), clonedBucket.getResourceId(), userRequest);
+    return StepResult.getStepResultSuccess();
+  }
+
+  private String randomBucketName() {
+    return UUID.randomUUID().toString().toLowerCase();
+  }
+
+  /**
+   * Build the list of IAM roles for this user. If the resource was initially shared, we make the
+   * cloned resource shared as well. If it's private, the user making the clone must be the owner
+   * and becomes EDITOR on the new resource.
+   *
+   * @param accessScope
+   * @return
+   */
+  private List<ControlledResourceIamRole> getIamRoles(AccessScopeType accessScope) {
```

Comment:
I'd like to be able to copy the user's old roles instead of just granting editor. I think that should be possible in Sam - I'm looking at `resourceRolesV2` in the client, though it isn't wrapped in SamService yet.

Reply:
That's different semantics than what I was thinking. I'm thinking a reader on a workspace or resource should be able to do whatever they want with their clone of that workspace/resource.